modules/core/src/main/java/org/apache/ignite/internal/client/thin/ReliableChannel.java

```diff
@@ -606,16 +603,271 @@ public void addChannelFailListener(Runnable chFailLsnr) {
+                newAddrs = parsedAddresses(hostAddrs);
+                prevHostAddrs = hostAddrs;
+            }
+        } else if (holders == null)
+            newAddrs = parsedAddresses(clientCfg.getAddresses());
+
+        if (newAddrs == null) {
+            finishChannelsReInit = System.currentTimeMillis();
+            return true;
+        }
+
+        Map<InetSocketAddress, ClientChannelHolder> curAddrs = new HashMap<>();
+        Set<InetSocketAddress> allAddrs = new HashSet<>(newAddrs.keySet());
+
+        if (holders != null) {
+            for (int i = 0; i < holders.size(); i++) {
+                ClientChannelHolder h = holders.get(i);
+
+                curAddrs.put(h.chCfg.getAddress(), h);
+                allAddrs.add(h.chCfg.getAddress());
+            }
+        }
+
+        List<ClientChannelHolder> reinitHolders = new ArrayList<>();
+
+        // The variable holds a new index of default channel after topology change.
+        // Suppose that reuse of the channel is better than open new connection.
+        int dfltChannelIdx = -1;
+
+        ClientChannelHolder currDfltHolder = null;
+
+        int idx = curChIdx;
+
+        if (idx != -1)
+            currDfltHolder = holders.get(idx);
+
+        for (InetSocketAddress addr : allAddrs) {
+            if (shouldStopChannelsReinit())
+                return false;
+
+            // Obsolete addr, to be removed.
+            if (!newAddrs.containsKey(addr)) {
+                curAddrs.get(addr).close();
+
+                continue;
+            }
+
+            // Create new holders for new addrs.
+            if (!curAddrs.containsKey(addr)) {
+                ClientChannelHolder hld = new ClientChannelHolder(new ClientChannelConfiguration(clientCfg, addr));
+
+                for (int i = 0; i < newAddrs.get(addr); i++)
+                    reinitHolders.add(hld);
+
+                continue;
+            }
+
+            // This holder is up to date.
+            ClientChannelHolder hld = curAddrs.get(addr);
+
+            for (int i = 0; i < newAddrs.get(addr); i++)
+                reinitHolders.add(hld);
+
+            if (hld == currDfltHolder)
+                dfltChannelIdx = reinitHolders.size() - 1;
+        }
+
+        if (dfltChannelIdx == -1)
+            dfltChannelIdx = new Random().nextInt(reinitHolders.size());
+
+        curChannelsGuard.writeLock().lock();
+        try {
+            channels = reinitHolders;
+            curChIdx = dfltChannelIdx;
+        }
+        finally {
+            curChannelsGuard.writeLock().unlock();
+        }
+
+        finishChannelsReInit = System.currentTimeMillis();
+        return true;
+    }
+
+    /**
+     * Establishing connections to servers. If partition awareness feature is enabled connections are created
+     * for every configured server. Otherwise only default channel is connected.
+     */
+    void channelsInit(boolean force) {
+        if (!force && channels != null)
```

Comment:
Currently, flag `force` is useless (at least for production code, but perhaps for test code too). There are 3 calls of this method from production code, first - right after constructor (when `channels == null`), two other calls with `force == true`.

Reply:
fixed